### DISCUSSION TITLE
RPM Reporting

### DISCUSSION THREAD

**Discussion by Cheesekeeper** [devcompl/TTController#116] [Q&A] [answered]
Hi, I just set up TT Controller on my system which is currently running a Floe DX 360 cooler, and 5 RiingPlus case fans.  Firstly, thank you so much to MoshiMoshi0 for developing this software.  The fans are virtually unusable with the official TT software, and I would have been tempted to replace the lot if it weren't for this.  I have successfully created a config (with help form these forums) that I am happy with and everything seems to be running well.

The only thing that's missing is a way to monitor fan speeds without running the service in console mode, or by rewiring the fans so the RPM values are visible to the motherboard and reported by HWInfo.  Was wondering if there is any way to extract the current RPM values from the service so that they can be displayed?  I use primarily RainMeter and HWInfo for system reporting (RainMeter easily reads values from HWInfo).  Rainmeter provides many ways to obtain info, so there would potentially be a lot of ways to get info from TTController into it.  One of the easiest would be if the service writes the current RPM values to the registry.  Could also read from a text file, although that would probably be inefficient.  RSS and web parsing is supported as well, so could get the values as part of a local HTTP request if the service supports it.  There is also a web socket plugin for RainMeter if it's possible to read values via the IPC extension.  Any thoughts on how to obtain these readings?  Ideally I'd like to be able to read from all fans, or at least two (one from the radiator, and one from the case fans).  Thanks very much.......

**Comment by MoshiMoshi0** [comment, marked as answer]
You would have to connect to the service websocket server and parse the received json in the rainmeter.
The service outputs data if you configure [IpcClientTimerInterval](https://moshimoshi0.github.io/TTController/config/service#ipcclienttimerinterval).

Note that im reworking the ipc stuff so it will probably break in the next update, whenever it will be released.

**Comment by Cheesekeeper** [comment]
Awesome, thank you :)  After reading this have been able to configure the IPC server and view the data in a web socket tester.  Now I need to figure out how to use the Rainmeter plugin to extract the values.  Did notice using the tester that if the client disconnects it makes the service crash.  Get the following when I terminate the connection:

```Unhandled Exception: System.Net.WebSockets.WebSocketException: An internal WebSocket error occurred. Please see the innerException, if present, for more details.  ---> System.Net.HttpListenerException: An operation was attempted on a nonexistent network connection
   at System.Net.WebSockets.WebSocketHttpListenerDuplexStream.WriteAsyncFast(HttpListenerAsyncEventArgs eventArgs)
   at System.Net.WebSockets.WebSocketHttpListenerDuplexStream.<MultipleWriteAsyncCore>d__38.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.WebSocketBase.<SendFrameAsync>d__48.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Net.WebSockets.WebSocketBase.WebSocketOperation.<Process>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.WebSocketBase.<SendAsyncCore>d__47.MoveNext()
   --- End of inner exception stack trace ---
   at System.Net.WebSockets.WebSocketBase.ThrowIfConvertibleException(String methodName, Exception exception, CancellationToken cancellationToken, Boolean aborted)
   at System.Net.WebSockets.WebSocketBase.<SendAsyncCore>d__47.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at TTController.Service.Ipc.WebSocketIpcServer.<SendStringAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at TTController.Service.Ipc.WebSocketIpcServer.<SendTask>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__6_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()```

**Reply by Cheesekeeper** [reply]
Couldn't find any way using existing modules to get the values into RainMeter so ended up creating my own service to read the values from the IPC server, extract the fan data, and write to the registry.  Will be interested to see what changes are in the next version of the IPC module, and will modify the service accordingly if it breaks something.  The exception mentioned above is still an issue - TTController crashes as soon as the listener service stops receiving, even if the socket connection is closed gracefully.

**Reply by MoshiMoshi0** [reply]
Yea I know about that exception but for me it happens very rarely.

You could also write your own IpcServer plugin instead of a service, but I will probably redesign most of ipc code.

**Reply by bonkas** [reply]
Is there a "for dummies" method to accomplish this? I would love to obtain the rpm data also but from the googling I have done on the IPC sockets stuff it has gone way more in depth than I want/understand even though I do have an IT background.

**Reply by Cheesekeeper** [reply]
Yeah I’d never touched the web sockets stuff before either, and it’s been quite a while since I have done any coding.  Mine’s not finished yet (works, but still needs some better error handling) but I’d be happy to send you the source code I have so far.  It’s in VB.Net.  Also haven't created a RainMeter skin for it yet, but with the data in the registry that should be really easy.

**Reply by bonkas** [reply]
Sure it couldnt hurt taking a look, might help me understand how it all works and try get some data out.

**Reply by Cheesekeeper** [reply]
Ok, source code of current version is attached.  Please note this is not finished - there is no logging yet, and still has debugging console output.  I haven't transferred it into a proper service yet either, but all the functionality is in place and seems to be working well as a console app.  It uses the Newtonsoft free JSON library to decode the output from the IPC service.  Paired with the updated version of the IPC library it successfully reconnects if the TTController service is restarted.  Also noticed that although I have used the command from MoshiMoshi0 to enable logging, it seems to be ignoring the interval sent in the command, and is using the interval defined in the config file instead.  Hope this proves useful....

[Module1.txt](https://github.com/MoshiMoshi0/TTController/files/5770518/Module1.txt)

**Comment by MoshiMoshi0** [comment]
Pushed initial changes of IPC stuff: https://ci.appveyor.com/project/MoshiMoshi0/ttcontroller/builds/37105992/artifacts

You now can enable/disable reporting of certain data by sending:
```{ "Enable": "<type>", "Interval": <update interval in ms> }```
```{ "Disable": "<type>" }```
to the service socket.

Where `<type>` is one of `"Colors", "Speed", "Data", "Config"`.
So for example by sending `{ "Enable": "Data", "Interval": 1000 }` you will receive port data every 1000ms.

You can also use multiple ipc plugins with the same `IpcName` (to enable/disable multiple triggers at the same time) and hopefully should not crash anymore.

**Reply by Cheesekeeper** [reply]
Nice :)  Confirmed no more crashing.  Updated my service to send the command to start reporting and it works flawlessly.  Thank you so much for your hard work on this...

**Reply by bonkas** [reply]
@Cheesekeeper I got my powershell script returning data from the websocket now, after looking at your code I notice the URL should be ws://x.x.x.x:port/service - I was missing /service.

@MoshiMoshi0 
Although after copying the latest changes I am getting the following errror when starting:

> 11:43:59.7810: ================================================================
> 11:43:59.7930: Initializing service, version "0.11.0-cleanup-ipc.1+31.Branch.feature-cleanup-ipc.Sha.aec6df923a9f29b404326ffadebdbf55e918ee5f"
> 11:44:00.0491: Loading plugin assembly: TTController.Plugin.DpsgController [0.11.0.0]
> 11:44:00.0491: Loading plugin assembly: TTController.Plugin.Level20DeskController [0.11.0.0]
> 11:44:00.0581: Loading plugin assembly: TTController.Plugin.RiingController [0.11.0.0]
> 11:44:00.0581: Loading plugin assembly: TTController.Plugin.RiingPlusController [0.11.0.0]
> 11:44:00.0581: Loading plugin assembly: TTController.Plugin.RiingQuadController [0.11.0.0]
> 11:44:00.0581: Loading plugin assembly: TTController.Plugin.RiingTrioController [0.11.0.0]
> 11:44:00.0581: Loading plugin assembly: TTController.Plugin.AuroraEffect [0.11.0.0]
> 11:44:00.0751: Loading plugin assembly: TTController.Plugin.BlendEffect [0.11.0.0]
> 11:44:00.0751: Loading plugin assembly: TTController.Plugin.BlinkEffect [0.11.0.0]
> 11:44:00.0751: Loading plugin assembly: TTController.Plugin.ColorSequenceEffect [0.11.0.0]
> 11:44:00.0751: Loading plugin assembly: TTController.Plugin.CopyColorEffect [0.11.0.0]
> 11:44:00.0891: Loading plugin assembly: TTController.Plugin.FlowEffect [0.11.0.0]
> 11:44:00.0891: Loading plugin assembly: TTController.Plugin.IpcEffect [0.11.0.0]
> 11:44:00.0891: Loading plugin assembly: TTController.Plugin.PingPongEffect [0.11.0.0]
> 11:44:00.0891: Loading plugin assembly: TTController.Plugin.PulseEffect [0.11.0.0]
> 11:44:00.0891: Loading plugin assembly: TTController.Plugin.RawEffect [0.11.0.0]
> 11:44:00.1070: Loading plugin assembly: TTController.Plugin.RazerConnectEffect [0.11.0.0]
> 11:44:00.1070: Loading plugin assembly: TTController.Plugin.RippleEffect [0.11.0.0]
> 11:44:00.1070: Loading plugin assembly: TTController.Plugin.SensorEffect [0.11.0.0]
> 11:44:00.1270: Loading plugin assembly: TTController.Plugin.SoundEffect [0.11.0.0]
> 11:44:00.1270: Loading plugin assembly: TTController.Plugin.SpectrumEffect [0.11.0.0]
> 11:44:00.1270: Loading plugin assembly: TTController.Plugin.StaticColorEffect [0.11.0.0]
> 11:44:00.1390: Loading plugin assembly: TTController.Plugin.WaveEffect [0.11.0.0]
> 11:44:00.1390: Loading plugin assembly: TTController.Plugin.ZoneEffect [0.11.0.0]
> 11:44:00.1390: Loading plugin assembly: TTController.Plugin.DelayPortModifier [0.11.0.0]
> Service failed with unhandled exception:
> System.TypeLoadException: Could not load type 'TTController.Common.Plugin.LedColorModifierBase`1' from assembly 'TTController.Common, Version=0.11.0.0, Culture=neutral, PublicKeyToken=null'.
>    at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
>    at System.Reflection.RuntimeAssembly.GetExportedTypes()
>    at TTController.Service.Utils.PluginLoader.<Search>d__6.MoveNext() in C:\projects\ttcontroller\Source\TTController.Service\Utils\PluginLoader.cs:line 59
>    at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
>    at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
>    at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
>    at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
>    at TTController.Service.Utils.PluginLoader.Load(String path, Type[] types) in C:\projects\ttcontroller\Source\TTController.Service\Utils\PluginLoader.cs:line 24
>    at TTController.Service.Utils.PluginLoader.LoadAll(String path) in C:\projects\ttcontroller\Source\TTController.Service\Utils\PluginLoader.cs:line 21
>    at TTController.Service.TTService.Initialize() in C:\projects\ttcontroller\Source\TTController.Service\TTService.cs:line 49
>    at TTController.Service.Program.<>c.<MainMenu>b__5_1() in C:\projects\ttcontroller\Source\TTController.Service\Program.cs:line 49

**Reply by MoshiMoshi0** [reply]
@bonkas remove everything but `config.json` and paste new release files, dont just copy and overwrite existing files

**Reply by bonkas** [reply]
@MoshiMoshi0 yup I did exactly this as you replied to my post and it fixed it. I was going to edit my post but you were way too fast to reply 👍 awesome stuff.
